src/core/components/navigator-stack/router.tsx

```diff
@@ -40,29 +52,63 @@ export const Router = (stackOptions: Types.StackOptions) => {
  */
 const mutateState = (
   state: StackNavigationState<Types.ParamListBase>,
+  stackOptions: Types.StackOptions,
   options: Types.RouterRenameOptions,
 ) => {
-  const routes = Array.from(state.routes);
+  const entrypointUrl = stackOptions.navContextProps?.entrypointUrl;
+  const doc = stackOptions.docContextProps?.getDoc();
+  const element = doc
+    ? NavigatorService.getNavigatorById(doc, stackOptions.id)
+    : null;
+  const elementRoutes = element
+    ? NavigatorService.getChildElements(element).filter(
+        e => e.localName === LOCAL_NAME.NAV_ROUTE,
+      )
+    : [];
 
-  const filteredNames = options.routeNames.filter((name: string) => {
-    return (
-      !options.routeKeyChanges?.includes(name) &&
-      !NavigatorService.isDynamicRoute(name)
-    );
+  const routes: Route[] = [];
+  const routeIds = state.routes.map((r: Route) => r.params?.id);
+  const routeHrefs = state.routes.map((r: Route) => {
+    return r.params?.url
+      ? NavigatorService.getUrlFromHref(r.params.url, entrypointUrl)
+      : undefined;
   });
 
-  filteredNames.forEach((name: string) => {
-    if (options.routeParamList) {
-      const params = options.routeParamList[name] || {};
-      if (!routes.find(route => route.name === name)) {
+  let sequenceBroken = false;
+  for (let i = 0; i < elementRoutes.length; i += 1) {
```

Comment:
the logic seems a bit intricate, would it be worth pulling it out in a helper method and write unit tests against it?

Reply:
Moved to 'helpers.ts', added tests.